### PR TITLE
RSDK-5316 - Add placeholder support for environment variables

### PR DESCRIPTION
**Pull request by zaporter-work**
Ticket: https://viam.atlassian.net/browse/RSDK-5316

### CODE REVIEW

File: config/placeholder_replacement.go [hunk 1]
```diff
@@ -23,6 +24,10 @@ var placeholderRegexp = regexp.MustCompile(`\$\{(?P<placeholder_key>[^\}]*)\}`)
 // packages.FutureP4ckge_Ty-pe.name.
 var packagePlaceholderRegexp = regexp.MustCompile(`^packages(\.(?P<type>[^\.]+))?\.(?P<name>[\w:/-]+)$`)
 
+// environmentPlaceholderRegexp matches on all valid ways of specifying one of our environment placeholders
+// This is compatible with IEEE Std 1003.1-2018 (see basedefs/V1_chap08.html).
+var environmentPlaceholderRegexp = regexp.MustCompile(`^environment\.(?P<name>[\w:/-]+)$`)
```
Comment (RoxyFarhad) [inline config/placeholder_replacement.go:29]: what was the decision ton use environment here and not {env.} I saw that the ticket indicates using environment but isn't that kind of against normal practice? 

Also have you checked that ${env.placeholder} for a local config doesn't break in this case either?

File: config/placeholder_replacement.go [hunk 2]
```diff
@@ -186,3 +199,17 @@ func (v *placeholderReplacementVisitor) replacePackagePlaceholder(toReplace stri
...
+
+func (v *placeholderReplacementVisitor) replaceEnvironmentPlaceholder(toReplace string) (string, error) {
+	matches := environmentPlaceholderRegexp.FindStringSubmatch(toReplace)
+	if matches == nil {
+		return toReplace, errors.Errorf("failed to find substring matches for %q", toReplace)
+	}
+	variableName := matches[environmentPlaceholderRegexp.SubexpIndex("name")]
+	value, present := os.LookupEnv(variableName)
+	if !present {
+		return toReplace, errors.Errorf("no environment variable named %q for placeholder %q",
```
Comment (RoxyFarhad) [inline config/placeholder_replacement.go:211]: super-nit: make named = defined

File: config/placeholder_replacement.go [hunk 3]
```diff
@@ -153,6 +158,14 @@ func (v *placeholderReplacementVisitor) replacePlaceholders(s string) (string, e
 			}
 			return []byte(replaced)
 		}
+		if environmentPlaceholderRegexp.Match(placeholderKey) {
```
Comment (RoxyFarhad) [inline config/placeholder_replacement.go:161]: could you break out the logic in this file into 2 separate helper functions? It looks messy and kind of hard to follow if you don't know what this doing.

File: config/placeholder_replacement.go (hunk 1 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:29]: > what was the decision ton use environment here and not {env.} I saw that the ticket indicates using environment but isn't that kind of against normal practice?

This came up in the scope and the consensus was that this gives us a lot more freedom going forward. If we just use ${env} we lock ourselves in. Ex: if there is an environment variable called `packages.ml_model.foo` what do we do?.  Or more specifically, if we decide to support other kinds placeholders in the future, we are at risk of breaking people's existing configs.

File: config/placeholder_replacement.go (hunk 2 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:211]: :+1:

File: config/placeholder_replacement.go (hunk 3 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:161]: Like another layer of nesting? Sure -- I can break it out a bit more -- not sure it will help readability but happy to do it.

File: config/placeholder_replacement.go (hunk 1 above)
Comment (RoxyFarhad) [reply config/placeholder_replacement.go:29]: >  Ex: if there is an environment variable called packages.ml_model.foo what do we do?

not sure what you mean by this with env vs. environment? 

Still not super sure how I follow the "locking ourselves in?"

File: config/placeholder_replacement.go (hunk 2 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:211]: So you're imagining `No environment variable defined HOME for placeholder ${environment.HOME}` instead of `No environment variable named HOME for placeholder ${environment.HOME}`

?

File: config/placeholder_replacement.go (hunk 1 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:29]: Oh you mean `env` vs `environment` not `${env}`

Comment (zaporter-work) [reply config/placeholder_replacement.go:29]: Umm, I don't think there was a strong reason one way or the other. This is just what people were happy with.

File: config/placeholder_replacement.go (hunk 3 above)
Comment (RoxyFarhad) [reply config/placeholder_replacement.go:161]: for example, break out the top function to be like matchOnPackagePlaceholder and this new logic could live in a separate file like matchOnEnvironmentPlaceholder? 

It will help readability because if you just look at those 2 function names its clear what the outer function is doing without having to read / understand a significant amount of code to do so. If you don't know how regex works in depth, this could be hard to follow.

File: config/placeholder_replacement.go (hunk 1 above)
Comment (RoxyFarhad) [reply config/placeholder_replacement.go:29]: sorry if I am taking this above the PR, but environment is just kind of wordy and there is a standard for this which is why I am wondering where the decision is to break the standard. Especially as we support env variables in local configs, so i feel like they should be synced while we do technically support both.

File: config/placeholder_replacement.go (hunk 2 above)
Comment (RoxyFarhad) [reply config/placeholder_replacement.go:211]: hm actually doesn't flow great. Its just that env variables store system variables, so naming just feels a bit wrong but its a super nit. 

Sorry just leave.

File: config/placeholder_replacement.go (hunk 3 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:161]: I'll try refactoring and see if I can get it looking a bit cleaner -- might try a switch statement.

File: config/placeholder_replacement.go [hunk 4]
```diff
@@ -144,18 +153,22 @@ func (v *placeholderReplacementVisitor) replacePlaceholders(s string) (string, e
...
-				return placeholder
-			}
-			return []byte(replaced)
+		switch {
+		case packagePlaceholderRegexp.Match(placeholderKey):
+			replacementResult, err = v.replacePackagePlaceholder(string(placeholderKey))
+		case environmentPlaceholderRegexp.Match(placeholderKey):
+			replacementResult, err = v.replaceEnvironmentPlaceholder(string(placeholderKey))
+		default:
+			err = errors.Errorf("invalid placeholder %q", string(placeholder))
```
Comment (zaporter-work) [inline config/placeholder_replacement.go:165]: I think this is much cleaner and should be easy for people to read / modify.

I tried your idea of splitting into a separate file, but the two files didn't make sense without each other but I don't think it ended up better -- the functions were very codependent.

File: config/placeholder_replacement.go (hunk 1 above)
Comment (zaporter-work) [reply config/placeholder_replacement.go:29]: Want to talk about this offline?

File: config/placeholder_replacement.go (hunk 4 above)
Comment (RoxyFarhad) [reply config/placeholder_replacement.go:165]: what was the dependency between functions? I am not super fussed on this because this definitely looks better but I just feel like having a readable function rather than the match code directly still would be better. Can refactor later as this placeholder stuff grows though.

**Review by RoxyFarhad — APPROVED** [review]
